### PR TITLE
NoiseMap Calibration Recipe Template and E2E Test

### PR DESCRIPTION
**Pull request by ell-bogat**
## Changes:
- Addresses the following issue: #147 
- Adds recipe templates for generating NoiseMap Calibration files, one for processing from L1 data, and one for processing from L2a data. Both may not be necessary, but I had to repeat all the processing from L1 anyway to have a direct comparison with the II&T pipeline, since their calibrate dark function operates directly on L1 data which must have a specific structure: The dataset must contain at least four "sub-stacks" of frames, with each stack having a unique combination of EM gain and exposure time. The L1 and L2a versions of these datasets, as well as the II&T processing result, are available in the TVAC box drive under TV-20_EXCAM_noise_characterization/noisemap_test_data.
  - The recipe template for processing from L1 includes `prescan_biassub` > `detect_cosmic_rays` > `correct_nonlinearity` > `update_to_l2a` > `save` > `convert_to_electrons` > `calibrate_darks` > `save`. 
  - The recipe template for processing from L2a includes only `convert_to_electrons` > `calibrate_darks` > `save`. 
  - We opted not to use the `select_frames` step to make sure each stack of frames with a unique EM gain & exposure time combination remained the same length.
  - We opted not to use the `desmear` step as none of the noise characterization images had an exposure time long enough for the noise to be dominated by dark current (which is the only noise source that is sensitive to the "smearing" effect. Kevin has opened an issue to allow the desmear step to intelligently apply itself only to the frames for which it is relevant.
  - We are also not currently using the `correct_bad_pixels` step, as this best recreates the II&T result, but we can add it in if it makes sense!
- Adds end-to-end tests for both recipes. The end-to-end tests compare the result from corgiDRP to the same dataset processed with the II&T pipeline and ensure that each science area pixel in each noise map (fixed-pattern noise, clock-induced charge, and dark current) differs by no more than 1E-5 between the corgiDRP and II&T pipeline. Currently the II&T result is saved as a fits file in the Box drive for reference, but I can provide the script I used to generate it if that is of interest.
  - As an example, here is the corgiDRP result compared to the II&T result for processing from L2a, with the difference (subtraction) of the two maps shown in the right hand panel for each noise map:
image
image
image

New files:
- corgidrp/recipe_templates/l1_to_l2a_noisemap.json : Recipe template for generating a NoiseMap calibration file from l1 data.
- corgidrp/recipe_templates/l2a_to_l2a_noisemap.json : Recipe template for generating a NoiseMap calibration file from l2a data.
- tests/e2e_tests/noisemap_cal_e2e.py : End-to-end tests for the NoiseMap recipe templates

Modified files:
- corgidrp/walker.py : Added calibrate_darks as a callable step
- corgidrp/darks.py : Allowed exposure time of 0 and EM gain of 1 in the files used for NoiseMap calibration.

## Type of change
- New feature (non-breaking change which adds functionality)

## Checklist before requesting a review
- [x] I have linted my code
- [x] I have verified that all unit tests pass in a clean environment and added new unit tests, as needed
- [x] I have verified that all docstrings are properly formatted and added new documentation, as needed

### CODE REVIEW

**Comment by ell-bogat** [comment]
Question for @semaphoreP : What's the best way to have GitHub run this e2e test on the right dataset and compare to the TVAC output? Right now on my laptop it takes like 20 minutes to run because I have it processing 192 files, which for this purpose it probably doesn't need to do, so I can make a smaller dataset and run the II&T code on that to get a result to compare to. But where should I host that data so GitHub can see it when running the test?

**Comment by ell-bogat** [comment]
Per our discussion at the CPP meeting today, here is an illustration of the row of nan pixels (in gray) that shows up at the top of the full frame detector image in the corgiDRP result. It's a PDF so you can zoom in and see the row I'm talking about:
[CorgiDRP_TVAC_Comparison_l2a_to_l2a_FPN_map_fullframe.pdf](https://github.com/user-attachments/files/16851102/CorgiDRP_TVAC_Comparison_l2a_to_l2a_FPN_map_fullframe.pdf)

**Comment by ell-bogat** [comment]
> Question for @semaphoreP : What's the best way to have GitHub run this e2e test on the right dataset and compare to the TVAC output? Right now on my laptop it takes like 20 minutes to run because I have it processing 192 files, which for this purpose it probably doesn't need to do, so I can make a smaller dataset and run the II&T code on that to get a result to compare to. But where should I host that data so GitHub can see it when running the test?

On this, I realized actually GitHub skips end to end tests when it does the automatic testing on a new pull request. I've just followed the example you wrote for the l1_to_l2a end to end test and uploaded the noisemap test datasets to the box drive. So if there's another way to best ensure reproduceability just let me know!

**Comment by ell-bogat** [comment]
There's a slightly related topic - Kevin had sent me some updated ii&t code, which is what I used for the comparison. Should I start a pull request to get that code onto the cgi_iit_drp github?

**Comment by semaphoreP** [comment]
> > Question for @semaphoreP : What's the best way to have GitHub run this e2e test on the right dataset and compare to the TVAC output? Right now on my laptop it takes like 20 minutes to run because I have it processing 192 files, which for this purpose it probably doesn't need to do, so I can make a smaller dataset and run the II&T code on that to get a result to compare to. But where should I host that data so GitHub can see it when running the test?
> 
> On this, I realized actually GitHub skips end to end tests when it does the automatic testing on a new pull request. I've just followed the example you wrote for the l1_to_l2a end to end test and uploaded the noisemap test datasets to the box drive. So if there's another way to best ensure reproduceability just let me know!

No need to make it faster. The goal of the e2e tests is to reproduce the actual use case. We will only run the e2e tests sparingly (either on pipeline release, or if you make a PR that affects a particular e2e test). So totally ok it takes 20 minutes. 

Test data on box sounds good!

File: corgidrp/darks.py [hunk 1]
```diff
@@ -527,14 +527,13 @@ def calibrate_darks_lsq(dataset, detector_params, detector_regions=None):
         raise CalDarksLSQException("Must have at least 2 unique EM gains "
                                    'represented by the sub-stacks in '
                                    'datasets.')
-    if len(EMgain_arr[EMgain_arr<=1]) != 0:
+    if len(EMgain_arr[EMgain_arr<=0]) != 0:
```
Comment (kjl0025) [inline corgidrp/darks.py:530]: Please change to `if len(EMgain_arr[EMgain_arr<1]) != 0:`

File: corgidrp/darks.py [hunk 2]
```diff
@@ -527,14 +527,13 @@ def calibrate_darks_lsq(dataset, detector_params, detector_regions=None):
...
         raise CalDarksLSQException('Each EM gain must be greater '
-            'than 1.')
+            'than 0.')
```
Comment (kjl0025) [inline corgidrp/darks.py:532]: Please change to `'Each EM gain must be 1 or greater.'`

**Comment by kjl0025** [comment]
> There's a slightly related topic - Kevin had sent me some updated ii&t code, which is what I used for the comparison. Should I start a pull request to get that code onto the cgi_iit_drp github?

Thanks, Ell.  I actually recently did a PR to get that code in there, so no worries.

**Review by maxwellmb — CHANGES_REQUESTED** [review]
Only one comment here: you've got to sort glob output - its unordered, so may vary between machines and OSes. You need to sort the output. Possible route recommended.

File: tests/e2e_tests/noisemap_cal_e2e.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    # make output directory if needed
+    noisemap_outputdir = os.path.join(e2eoutput_path, "noisemap_output")
+    if not os.path.exists(noisemap_outputdir):
+        os.mkdir(noisemap_outputdir)
+
+    # assume all cals are in the same directory
+    nonlin_path = os.path.join(processed_cal_path, "nonlin_table_240322.txt")
+    
+    # define the raw science data to process
+    l1_data_filelist = glob(os.path.join(l1_datadir,"*.fits"))
```
[the remark this answers is not in the document]
Comment (ell-bogat) [reply tests/e2e_tests/noisemap_cal_e2e.py:32]: Gotcha, thanks!

File: tests/e2e_tests/noisemap_cal_e2e.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+    # Nonlinearity calibration
+    nonlin_dat = np.genfromtxt(nonlin_path, delimiter=",")
+    nonlinear_cal = data.NonLinearityCalibration(nonlin_dat, pri_hdr=pri_hdr, ext_hdr=ext_hdr,
+                                                input_dataset=mock_input_dataset)
+    nonlinear_cal.save(filedir=noisemap_outputdir, filename="mock_nonlinearcal.fits" )
+    this_caldb.create_entry(nonlinear_cal)
+
+    # KGain
+    kgain_val = 8.7
```
[the remark this answers is not in the document]
Comment (ell-bogat) [reply tests/e2e_tests/noisemap_cal_e2e.py]: Just added a comment to this line (and the equivalent line in the second test) that the 8.7 number comes from the TVAC-20 noise characterization measurements. @maxwellmb let me know if that satisfies this request and then I think the PR is ready to merge!

**Comment by maxwellmb** [comment]
It seems as if the filename that is picked to generate the NoiseMaps calibration changes from run to run (at least on my machine). 
I ran the tests twice and got a different calibration file in each run: 
`CGI_EXCAM_L2a_0000050772_DetectorNoiseMaps.fits` and then `CGI_EXCAM_L2a_0000064235_DetectorNoiseMaps.fits `. This throws an error because it doesn't match the hardcoded expected filename.

**Review by maxwellmb — APPROVED** [review]
All done! Thanks @ell-bogat

**Pull request merged**